# HW 4 - Neural POS Tagger

In this exercise, you are going to build a set of deep learning models on part-of-speech (POS) tagging using Tensorflow and Keras. Tensorflow is a deep learning framwork developed by Google, and Keras is a frontend library built on top of Tensorflow (or Theano, CNTK) to provide an easier way to use standard layers and networks.

To complete this exercise, you will need to build deep learning models for POS tagging in Thai using NECTEC's ORCHID corpus. You will build one model for each of the following type:

- Neural POS Tagging with Word Embedding using Fixed / non-Fixed Pretrained weights
- Neural POS Tagging with Viterbi / Marginal CRF

Pretrained word embeddding are already given for you to use (albeit, a very bad one). Optionally, you can use your best pretrained word embeddding from previous exercise.

We also provide the code for data cleaning, preprocessing and some starter code for keras in this notebook but feel free to modify those parts to suit your needs. You can also complete this exercise using only Tensorflow (without using Keras). Feel free to use additional libraries (e.g. scikit-learn) as long as you have a model for each type mentioned above.

### Don't forget to shut down your instance on Gcloud when you are not using it ###

## 1. Setup and Preprocessing

We use POS data from [ORCHID corpus](https://www.nectec.or.th/corpus/index.php?league=pm), which is a POS corpus for Thai language.
A method used to read the corpus into a list of sentences with (word, POS) pairs have been implemented already. The example usage has shown below.
We also create a word vector for unknown word by random.

In [1]:
from data.orchid_corpus import get_sentences
import numpy as np
import numpy.random
import keras.preprocessing
np.random.seed(42)

Using TensorFlow backend.


In [2]:
unk_emb =np.random.randn(32)
train_data = get_sentences('train')
test_data = get_sentences('test')
print(train_data[0])

[('การ', 'FIXN'), ('ประชุม', 'VACT'), ('ทาง', 'NCMN'), ('วิชาการ', 'NCMN'), ('<space>', 'PUNC'), ('ครั้ง', 'CFQC'), ('ที่ 1', 'DONM')]


Next, we load pretrained weight embedding using pickle. The pretrained weight is a dictionary which map a word to its embedding.

In [3]:
import pickle
fp = open('basic_ff_embedding.pt', 'rb')
embeddings = pickle.load(fp)
fp.close()

The given code below generates an indexed dataset(each word is represented by a number) for training and testing data. The index 0 is reserved for padding to help with variable length sequence. (Additionally, You can read more about padding here [https://machinelearningmastery.com/data-preparation-variable-length-input-sequences-sequence-prediction/])

## 2. Prepare Data

In [4]:
word_to_idx ={}
idx_to_word ={}
label_to_idx = {}
for sentence in train_data:
    for word,pos in sentence:
        if word not in word_to_idx:
            word_to_idx[word] = len(word_to_idx)+1
            idx_to_word[word_to_idx[word]] = word
        if pos not in label_to_idx:
            label_to_idx[pos] = len(label_to_idx)+1
word_to_idx['UNK'] = len(word_to_idx)

n_classes = len(label_to_idx.keys())+1

This section is tweaked a little from the demo, word2features will return word index instead of features, and sent2labels will return a sequence of word indices in the sentence.

In [5]:
def word2features(sent, i, emb):
    word = sent[i][0]
    if word in word_to_idx :
        return word_to_idx[word]
    else :
        return word_to_idx['UNK']

def sent2features(sent, emb_dict):
    return np.asarray([word2features(sent, i, emb_dict) for i in range(len(sent))])

def sent2labels(sent):
    return numpy.asarray([label_to_idx[label] for (word, label) in sent],dtype='int32')

def sent2tokens(sent):
    return [word for (word, label) in sent]

In [6]:
sent2features(train_data[100], embeddings)

array([ 29, 327,   5, 328])

Next we create train and test dataset, then we use keras to post-pad the sequence to max sequence with 0. Our labels are changed to a one-hot vector.

In [7]:
%%time
x_train = np.asarray([sent2features(sent, embeddings) for sent in train_data])
y_train = [sent2labels(sent) for sent in train_data]
x_test = [sent2features(sent, embeddings) for sent in test_data]
y_test = [sent2labels(sent) for sent in test_data]

CPU times: user 362 ms, sys: 15.3 ms, total: 377 ms
Wall time: 376 ms


In [8]:
x_train=keras.preprocessing.sequence.pad_sequences(x_train, maxlen=None, dtype='int32', padding='post', truncating='pre', value=0.)
y_train=keras.preprocessing.sequence.pad_sequences(y_train, maxlen=None, dtype='int32', padding='post', truncating='pre', value=0.)
x_test=keras.preprocessing.sequence.pad_sequences(x_test, maxlen=102, dtype='int32', padding='post', truncating='pre', value=0.)
y_temp =[]
for i in range(len(y_train)):
    y_temp.append(np.eye(n_classes)[y_train[i]][np.newaxis,:])
y_train = np.asarray(y_temp).reshape(-1,102,n_classes)
del(y_temp)

In [9]:
print(x_train[100],x_train.shape)
print(y_train[100][3],y_train.shape)

[ 29 327   5 328   0   0   0   0   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0   0   0   0   0   0   0] (18500, 102)
[ 0.  0.  0.  1.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.
  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.
  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.  0.] (18500, 102, 48)


## 3. Evaluate

Our output from keras is a distribution of problabilities on all possible label. outputToLabel will return an indices of maximum problability from output sequence.

evaluation_report is the same as in the demo

In [10]:
def outputToLabel(yt,seq_len):
    out = []
    for i in range(0,len(yt)):
        if(i==seq_len):
            break
        out.append(np.argmax(yt[i]))
    return out

In [11]:
import pandas as pd
from IPython.display import display

def evaluation_report(y_true, y_pred):
    # retrieve all tags in y_true
    tag_set = set()
    for sent in y_true:
        for tag in sent:
            tag_set.add(tag)
    for sent in y_pred:
        for tag in sent:
            tag_set.add(tag)
    tag_list = sorted(list(tag_set))
    
    # count correct points
    tag_info = dict()
    for tag in tag_list:
        tag_info[tag] = {'correct_tagged': 0, 'y_true': 0, 'y_pred': 0}

    all_correct = 0
    all_count = sum([len(sent) for sent in y_true])
    for sent_true, sent_pred in zip(y_true, y_pred):
        for tag_true, tag_pred in zip(sent_true, sent_pred):
            if tag_true == tag_pred:
                tag_info[tag_true]['correct_tagged'] += 1
                all_correct += 1
            tag_info[tag_true]['y_true'] += 1
            tag_info[tag_pred]['y_pred'] += 1
    accuracy = (all_correct / all_count) * 100
            
    # summarize and make evaluation result
    eval_list = list()
    for tag in tag_list:
        eval_result = dict()
        eval_result['tag'] = tag
        eval_result['correct_count'] = tag_info[tag]['correct_tagged']
        precision = (tag_info[tag]['correct_tagged']/tag_info[tag]['y_pred'])*100 if tag_info[tag]['y_pred'] else '-'
        recall = (tag_info[tag]['correct_tagged']/tag_info[tag]['y_true'])*100 if (tag_info[tag]['y_true'] > 0) else 0
        eval_result['precision'] = precision
        eval_result['recall'] = recall
        eval_result['f_score'] = (2*precision*recall)/(precision+recall) if (type(precision) is float and recall > 0) else '-'
        
        eval_list.append(eval_result)

    eval_list.append({'tag': 'accuracy=%.2f' % accuracy, 'correct_count': '', 'precision': '', 'recall': '', 'f_score': ''})
    
    df = pd.DataFrame.from_dict(eval_list)
    df = df[['tag', 'precision', 'recall', 'f_score', 'correct_count']]
    display(df)

## 4. Train a model

In [12]:
from keras.models import Sequential, Model
from keras.layers import Embedding, Reshape, Activation, Input, Dense,GRU,Reshape,TimeDistributed,Bidirectional,Dropout,Masking
from keras_contrib.layers import CRF
from keras.optimizers import Adam

The model is this section is separated to two groups

- Neural POS Tagger (4.1)
- Neural CRF POS Tagger (4.2)

## 4.1.1 Neural POS Tagger  (Example)

We create a simple Neural POS Tagger as an example for you. This model dosen't use any pretrained word embbeding so it need to use Embedding layer to train the word embedding from scratch.

In [54]:
model = Sequential()
model.add(Embedding(len(word_to_idx), 32, input_length=102, mask_zero=True, weights=[np.array(pre_em)], trainable=False))
model.add(Bidirectional(GRU(32, return_sequences=True)))
model.add(Dropout(0.2))
model.add(TimeDistributed(Dense(n_classes,activation='softmax')))
model.summary()
adam  = Adam(lr=0.001)
model.compile(optimizer=adam,  loss='categorical_crossentropy', metrics=['categorical_accuracy'])

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_2 (Embedding)      (None, 102, 32)           480608    
_________________________________________________________________
bidirectional_2 (Bidirection (None, 102, 64)           12480     
_________________________________________________________________
dropout_2 (Dropout)          (None, 102, 64)           0         
_________________________________________________________________
time_distributed_2 (TimeDist (None, 102, 48)           3120      
Total params: 496,208
Trainable params: 15,600
Non-trainable params: 480,608
_________________________________________________________________


In [56]:
%%time
model.fit(x_train,y_train,batch_size=128,epochs=10,verbose=1)

Epoch 1/10
18500/18500 [==============================] - 61s - loss: 2.4416 - categorical_accuracy: 0.3045    
Epoch 2/10
18500/18500 [==============================] - 61s - loss: 2.1849 - categorical_accuracy: 0.3392    
Epoch 3/10
18500/18500 [==============================] - 61s - loss: 2.1038 - categorical_accuracy: 0.3610    
Epoch 4/10
18500/18500 [==============================] - 61s - loss: 2.0235 - categorical_accuracy: 0.3910    
Epoch 5/10
18500/18500 [==============================] - 61s - loss: 1.9109 - categorical_accuracy: 0.4411    
Epoch 6/10
18500/18500 [==============================] - 61s - loss: 1.7884 - categorical_accuracy: 0.4844    
Epoch 7/10
18500/18500 [==============================] - 61s - loss: 1.6992 - categorical_accuracy: 0.5110    
Epoch 8/10
18500/18500 [==============================] - 61s - loss: 1.6319 - categorical_accuracy: 0.5318    
Epoch 9/10
18500/18500 [==============================] - 61s - loss: 1.5760 - categorical_accuracy: 0.5

In [57]:
%%time
model.save_weights('/data/my_pos_no_crf.h5')
# model.load_weights('/data/my_pos_no_crf.h5')
y_pred=model.predict(x_test)
ypred = [outputToLabel(y_pred[i],len(y_test[i])) for i in range(len(y_pred))]
evaluation_report(y_test, ypred)

tag precision    recall   f_score correct_count
0                1   93.2294   99.0231    96.039          3649
1                2   54.6838   59.2386   56.8702          4886
2                3   49.3192   60.0509   54.1585         10142
3                4   55.8613   87.9895   68.3375         11370
4                5         -         0         -             0
5                6         0         0         -             0
6                7   93.0233   76.9601   84.2327          1600
7                8   84.2105   3.85542   7.37327            16
8                9         -         0         -             0
9               10         -         0         -             0
10              11         -         0         -             0
11              12   79.1086   35.4557   48.9655           284
12              13   49.9587   33.6672   40.2261          1210
13              14   67.1845   70.1221   68.6219          3849
14              15   43.3333   1.15865   2.25694            13
15              16   51.1358   40.2998   45.0757           968
16              17         0         0         -             0
17              18   62.1779   64.7619   63.4436           748
18              19         -         0         -             0
19              20         -         0         -             0
20              21   71.5517   32.8713   45.0475           498
21              22   46.6667  0.436681  0.865266             7
22              23   45.5336   40.8221   43.0493           576
23              24   69.2982   34.5355   46.0977           316
24              25         -         0         -             0
25              26         -         0         -             0
26              27         -         0         -             0
27              29         -         0         -             0
28              30         -         0         -             0
29              31         -         0         -             0
30              32         -         0         -             0
31              33         -         0         -             0
32              34   82.4053   65.8363   73.1949           370
33              35         -         0         -             0
34              36         -         0         -             0
35              37         -         0         -             0
36              38         -         0         -             0
37              39         -         0         -             0
38              40         -         0         -             0
39              41         -         0         -             0
40              42         -         0         -             0
41              43         -         0         -             0
42              45         -         0         -             0
43              46         -         0         -             0
44  accuracy=57.83

CPU times: user 44.2 s, sys: 7.43 s, total: 51.6 s
Wall time: 17.9 s


## 4.1.2 Neural POS Tagger - Fix Weight

### #TODO 1
We would like you create a neural postagger model with keras with the pretrained word embedding as an input. The word embedding should be fixed across training time. To finish this excercise you must train the model and show the evaluation report with this model as shown in the example.

(You may want to read about Keras's Masking layer)

Optionally, you can use your own pretrained word embedding from previous homework

In [58]:
%%time
# Write your code here
pre_em = []
pre_em.append(np.zeros(32))
for i in range(1,len(idx_to_word)+1):
    if(idx_to_word[i] in embeddings.keys()):
        pre_em.append(embeddings[idx_to_word[i]])
    else:
        pre_em.append(np.zeros(32))

CPU times: user 28.3 ms, sys: 170 µs, total: 28.4 ms
Wall time: 27.7 ms


## 4.1.3 Neural POS Tagger - Trainable pretrained weight

### #TODO 2
We would like you create a neural postagger model with keras with the pretrained word embedding as an input. However The word embedding is trainable (not fixed). To finish this excercise you must train the model and show the evaluation report with this model as shown in the example.

Please note that the given pretrained word embedding only have weights for the vocabuary in BEST corpus from previous homework.

Optionally, you can use your own pretrained word embedding from previous homework.

In [59]:
# Write your code here
model = Sequential()
model.add(Embedding(len(word_to_idx), 32, input_length=102, mask_zero=True, weights=[np.array(pre_em)], trainable=True))
model.add(Bidirectional(GRU(32, return_sequences=True)))
model.add(Dropout(0.2))
model.add(TimeDistributed(Dense(n_classes,activation='softmax')))
model.summary()
adam  = Adam(lr=0.001)
model.compile(optimizer=adam,  loss='categorical_crossentropy', metrics=['categorical_accuracy'])

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_3 (Embedding)      (None, 102, 32)           480608    
_________________________________________________________________
bidirectional_3 (Bidirection (None, 102, 64)           12480     
_________________________________________________________________
dropout_3 (Dropout)          (None, 102, 64)           0         
_________________________________________________________________
time_distributed_3 (TimeDist (None, 102, 48)           3120      
Total params: 496,208
Trainable params: 496,208
Non-trainable params: 0
_________________________________________________________________


In [61]:
%%time
model.fit(x_train,y_train,batch_size=256,epochs=10,verbose=1)
model.save_weights('/data/my_pos_w_crf.h5')

Epoch 1/10
18500/18500 [==============================] - 33s - loss: 1.5628 - categorical_accuracy: 0.6395    
Epoch 2/10
18500/18500 [==============================] - 33s - loss: 0.9794 - categorical_accuracy: 0.7935    
Epoch 3/10
18500/18500 [==============================] - 33s - loss: 0.6562 - categorical_accuracy: 0.8591    
Epoch 4/10
18500/18500 [==============================] - 33s - loss: 0.4895 - categorical_accuracy: 0.8904    
Epoch 5/10
18500/18500 [==============================] - 33s - loss: 0.3918 - categorical_accuracy: 0.9103    
Epoch 6/10
18500/18500 [==============================] - 33s - loss: 0.3274 - categorical_accuracy: 0.9242    
Epoch 7/10
18500/18500 [==============================] - 33s - loss: 0.2837 - categorical_accuracy: 0.9330    
Epoch 8/10
18500/18500 [==============================] - 34s - loss: 0.2533 - categorical_accuracy: 0.9382    
Epoch 9/10
18500/18500 [==============================] - 33s - loss: 0.2313 - categorical_accuracy: 0.9

In [62]:
%%time
model.load_weights('/data/my_pos_w_crf.h5')
y_pred=model.predict(x_test)
ypred = [outputToLabel(y_pred[i],len(y_test[i])) for i in range(len(y_pred))]
evaluation_report(y_test, ypred)

tag precision   recall  f_score correct_count
0                1   99.8368  99.5929  99.7147          3670
1                2   95.0721  93.5621   94.311          7717
2                3   89.6411     97.3  93.3136         16433
3                4   99.7134  99.6363  99.6748         12875
4                5   95.2381  89.5522  92.3077            60
5                6   98.0603  87.1648  92.2921           455
6                7   97.1278   97.595  97.3608          2029
7                8        75  33.9759  46.7662           141
8                9   75.8621  47.8261  58.6667           176
9               10   60.5787  39.9285  48.1322           335
10              11   89.8876  93.0233  91.4286            80
11              12   96.1776  97.3783  96.7742           780
12              13   87.8676  86.4496  87.1529          3107
13              14   94.5606  93.1135  93.8315          5111
14              15   80.9899  64.1711  71.6062           720
15              16   87.1255   89.592  88.3415          2152
16              17   97.6744  86.3014  91.6364           504
17              18   97.3662  99.2208  98.2847          1146
18              19   97.9104  95.0725  96.4706           328
19              20   98.9619  96.9492  97.9452           286
20              21   92.4142  92.4752  92.4447          1401
21              22   81.8302  77.5421  79.6284          1243
22              23   87.6375  95.9603  91.6103          1354
23              24   80.7018  85.4645  83.0149           782
24              25    91.958  63.6804  75.2504           263
25              26   87.6543  80.6818  84.0237           142
26              27   92.3077  82.4427  87.0968           108
27              29   95.0311  96.8354  95.9248           306
28              30        75  70.5882  72.7273            72
29              31   53.9773   92.233  68.1004            95
30              32   95.2381  33.7079  49.7925            60
31              33         -        0        -             0
32              34   88.1956  93.0605  90.5628           523
33              35         -        0        -             0
34              36         -        0        -             0
35              37    90.991  99.0196  94.8357           101
36              38       100   2.5641        5             1
37              39   78.2609  77.1429  77.6978           108
38              40       100  99.2857  99.6416           278
39              41         -        0        -             0
40              42       100  17.6471       30             3
41              43         -        0        -             0
42              45         -        0        -             0
43              46         -        0        -             0
44  accuracy=92.73

CPU times: user 44.6 s, sys: 6.85 s, total: 51.4 s
Wall time: 17.9 s


### #TODO 3
Compare the result between all neural tagger models in 4.1.x and provide a convincing reason and example for the result of these models (which model perform best or worst, why?)

(If you use your own weight please state so in the answer)

<b>Write your answer here :</b>
<pre style="background-color: lightgreen">
Since pre-trained weight has already been set up to better initial condition from pre-trained words, it performed better in accuracy with 92.73 to 57.83 accuracy. Also the non-pretrained has set `Trainable` to `False`, so the weight is fixed at random vaule. Eventhough the overall network perform better as more epoch, the model cannot be changed. Thus lead to the lower accuracy rate.
</pre>

## 4.2.1 CRF Viterbi

Your next two tasks are to incorporate Conditional random fields (CRF) to your model. <b>You do not need to use pretrained weight</b>.

Keras already implement a CRF neural model for you. However, you need to use the official extension repository for Keras library, call keras-contrib. You should read about keras-contrib crf layer before attempt this exercise section

### #TODO 4
Use Keras-contrib CRF layer in your model. You should set the layer parameter so it can give the best performance on testing using <b>viterbi algorithm</b>. Your model must use crf for loss function and metric. CRF is quite complex compare to previous example model, so you should train it with more epoch, so it can converge.

To finish this excercise you must train the model and show the evaluation report with this model as shown in the example.

Do not forget to save this model weight.

In [23]:
# Write your code here
from keras_contrib.layers import CRF
from keras.callbacks import ReduceLROnPlateau
from keras import regularizers

In [79]:
# Write your code here
import keras.backend as K
K.clear_session()
model = Sequential()
model.add(Embedding(len(word_to_idx),32,input_length=102,mask_zero=True))
model.add(Bidirectional(GRU(32, return_sequences=True)))
model.add(Dropout(0.2))
model.add(TimeDistributed(Dense(n_classes, activation = 'relu')))
crf = CRF(n_classes,
          learn_mode='join',
          test_mode='viterbi',
          sparse_target=False,
          use_boundary=True,
          use_bias=True,
          activation='linear',
          kernel_initializer='glorot_uniform',
          chain_initializer='orthogonal',
          bias_initializer='zeros',
          boundary_initializer='zeros',
          kernel_regularizer=None,
          chain_regularizer=None,
          boundary_regularizer=None,
          bias_regularizer=None,
          kernel_constraint=None,
          chain_constraint=None,
          boundary_constraint=None,
          bias_constraint=None,
          input_dim=None,
          unroll=False)
model.add(crf)
model.summary()
adam  = Adam(lr=0.0015)
model.compile(optimizer=adam,loss=crf.loss_function, metrics=[crf.accuracy])

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, 102, 32)           480608    
_________________________________________________________________
bidirectional_1 (Bidirection (None, 102, 64)           12480     
_________________________________________________________________
dropout_1 (Dropout)          (None, 102, 64)           0         
_________________________________________________________________
time_distributed_1 (TimeDist (None, 102, 48)           3120      
_________________________________________________________________
crf_1 (CRF)                  (None, 102, 48)           4752      
Total params: 500,960
Trainable params: 500,960
Non-trainable params: 0
_________________________________________________________________


In [42]:
%%time
reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.05, patience=3, min_lr=0.001)
model.fit(x_train,y_train,batch_size=128,epochs=5,verbose=1,shuffle=True,validation_split=0.2,
         callbacks=[reduce_lr])

Train on 14800 samples, validate on 3700 samples
Epoch 1/5
14800/14800 [==============================] - 71s - loss: 46.9092 - acc: 0.4141 - val_loss: 44.2042 - val_acc: 0.7787
Epoch 2/5
14800/14800 [==============================] - 71s - loss: 45.1785 - acc: 0.8762 - val_loss: 43.6239 - val_acc: 0.8971
Epoch 3/5
14800/14800 [==============================] - 71s - loss: 44.9055 - acc: 0.9289 - val_loss: 43.5473 - val_acc: 0.9093
Epoch 4/5
14800/14800 [==============================] - 70s - loss: 44.8405 - acc: 0.9409 - val_loss: 43.5197 - val_acc: 0.9157
Epoch 5/5
14800/14800 [==============================] - 70s - loss: 44.8113 - acc: 0.9469 - val_loss: 43.5070 - val_acc: 0.9172
CPU times: user 15min 57s, sys: 2min 17s, total: 18min 15s
Wall time: 5min 57s


In [43]:
y_pred=model.predict(x_test)
ypred = [outputToLabel(y_pred[i],len(y_test[i])) for i in range(len(y_pred))]
evaluation_report(y_test, ypred)

tag precision   recall  f_score correct_count
0                1   99.8911  99.5929  99.7418          3670
1                2   91.9514  94.6048  93.2592          7803
2                3   91.1332  94.5704    92.82         15972
3                4   99.9922  99.5976  99.7945         12870
4                5   86.8421  98.5075  92.3077            66
5                6   98.6957  86.9732  92.4644           454
6                7    98.105   97.114   97.607          2019
7                8   71.6049  27.9518   40.208           116
8                9   73.8351  55.9783  63.6785           206
9               10   61.1111  40.6436  48.8189           341
10              11   92.0455   94.186  93.1034            81
11              12   93.3254  97.7528  95.4878           783
12              13   88.1933  85.8375  86.9994          3085
13              14   94.2868   94.407  94.3468          5182
14              15   64.7569  66.4884  65.6113           746
15              16   86.3019  89.9667  88.0962          2161
16              17   97.8641  86.3014  91.7197           504
17              18   96.7089  99.2208  97.9487          1146
18              19   97.9104  95.0725  96.4706           328
19              20   98.9619  96.9492  97.9452           286
20              21   95.2876  90.7591  92.9682          1375
21              22   79.8949  75.8578   77.824          1216
22              23   88.8089  96.1729  92.3443          1357
23              24   79.6488  84.2623  81.8906           771
24              25   86.0058  71.4286  78.0423           295
25              26   88.8268  90.3409  89.5775           159
26              27   88.1356  79.3893  83.5341           104
27              29   92.7492  97.1519  94.8995           307
28              30    78.125  73.5294  75.7576            75
29              31   56.6879  86.4078  68.4615            89
30              32   95.5224  35.9551  52.2449            64
31              33         -        0        -             0
32              34   81.1321  91.8149  86.1436           516
33              35   33.3333  11.1111  16.6667             1
34              36         -        0        -             0
35              37   91.8182  99.0196   95.283           101
36              38   58.6207  43.5897       50            17
37              39   73.8372  90.7143  81.4103           127
38              40       100  99.6429  99.8211           279
39              41       100       15   26.087             3
40              42       100  17.6471       30             3
41              43         -        0        -             0
42              45         -        0        -             0
43              46         -        0        -             0
44  accuracy=92.35

In [44]:
model.fit(x_train,y_train,batch_size=128,epochs=5,verbose=1,shuffle=True,validation_split=0.2, callbacks=[reduce_lr])

Train on 14800 samples, validate on 3700 samples
Epoch 1/5
14800/14800 [==============================] - 70s - loss: 44.7940 - acc: 0.9499 - val_loss: 43.4974 - val_acc: 0.9192
Epoch 2/5
14800/14800 [==============================] - 70s - loss: 44.7811 - acc: 0.9534 - val_loss: 43.4961 - val_acc: 0.9195
Epoch 3/5
14800/14800 [==============================] - 70s - loss: 44.7715 - acc: 0.9553 - val_loss: 43.4959 - val_acc: 0.9212
Epoch 4/5
14800/14800 [==============================] - 70s - loss: 44.7635 - acc: 0.9572 - val_loss: 43.4973 - val_acc: 0.9223
Epoch 5/5
14800/14800 [==============================] - 70s - loss: 44.7558 - acc: 0.9592 - val_loss: 43.4970 - val_acc: 0.9213


In [45]:
y_pred=model.predict(x_test)
ypred = [outputToLabel(y_pred[i],len(y_test[i])) for i in range(len(y_pred))]
evaluation_report(y_test, ypred)

tag precision   recall  f_score correct_count
0                1   99.8639  99.5929  99.7283          3670
1                2   94.6768  93.5863  94.1284          7719
2                3   91.1388  95.6717  93.3503         16158
3                4   99.9534  99.6518  99.8024         12877
4                5   91.6667  98.5075   94.964            66
5                6   99.1266  86.9732  92.6531           454
6                7   98.2885  96.6811  97.4782          2010
7                8   72.1992  41.9277  53.0488           174
8                9   70.8738  59.5109  64.6972           219
9               10   62.1572  40.5244   49.062           340
10              11   91.3043  97.6744   94.382            84
11              12   93.2384  98.1273  95.6204           786
12              13   85.8514  87.1174  86.4798          3131
13              14   94.8402  94.0973  94.4673          5165
14              15   70.9282  72.1925  71.5548           810
15              16   87.0865  90.9659  88.9839          2185
16              17   96.6605  89.2123  92.7872           521
17              18   96.3087  99.3939   97.827          1148
18              19   97.3607  96.2319   96.793           332
19              20   98.6207  96.9492  97.7778           286
20              21   93.1864  92.0792  92.6295          1395
21              22   84.6154   74.111  79.0156          1188
22              23   90.2276  95.5351  92.8055          1348
23              24   85.0273  85.0273  85.0273           778
24              25   89.2537  72.3971  79.9465           299
25              26   93.6047  91.4773  92.5287           161
26              27   95.9184  71.7557  82.0961            94
27              29   93.4169  94.3038  93.8583           298
28              30    77.551  74.5098       76            76
29              31   58.3333  81.5534  68.0162            84
30              32   65.0602  60.6742  62.7907           108
31              33         -        0        -             0
32              34   87.3264  89.5018  88.4007           503
33              35   83.3333  55.5556  66.6667             5
34              36       100     87.5  93.3333            14
35              37    90.991  99.0196  94.8357           101
36              38     56.25  46.1538  50.7042            18
37              39   69.9482  96.4286  81.0811           135
38              40       100      100      100           280
39              41        75       75       75            15
40              42       100  64.7059  78.5714            11
41              43         -        0        -             0
42              45         -        0        -             0
43              46         -        0        -             0
44  accuracy=92.87

In [53]:
model.save_weights('/data/viterbi_crf_2.h5')

In [75]:
model.load_weights('/data/viterbi_crf.h5')

In [80]:
%%time
reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.05, patience=3, min_lr=0.001)
model.fit(x_train,y_train,batch_size=256,epochs=10,verbose=1,shuffle=True,validation_split=0.2, callbacks=[reduce_lr])

Train on 14800 samples, validate on 3700 samples
Epoch 1/10
14800/14800 [==============================] - 39s - loss: 47.6976 - acc: 0.2141 - val_loss: 45.3184 - val_acc: 0.4184
Epoch 2/10
14800/14800 [==============================] - 38s - loss: 46.0685 - acc: 0.6046 - val_loss: 44.0675 - val_acc: 0.8035
Epoch 3/10
14800/14800 [==============================] - 37s - loss: 45.1702 - acc: 0.8771 - val_loss: 43.6798 - val_acc: 0.8888
Epoch 4/10
14800/14800 [==============================] - 38s - loss: 44.9468 - acc: 0.9220 - val_loss: 43.5998 - val_acc: 0.9045
Epoch 5/10
14800/14800 [==============================] - 38s - loss: 44.8758 - acc: 0.9358 - val_loss: 43.5673 - val_acc: 0.9097
Epoch 6/10
14800/14800 [==============================] - 38s - loss: 44.8432 - acc: 0.9415 - val_loss: 43.5495 - val_acc: 0.9120
Epoch 7/10
14800/14800 [==============================] - 38s - loss: 44.8228 - acc: 0.9446 - val_loss: 43.5406 - val_acc: 0.9134
Epoch 8/10
14800/14800 [=================

In [81]:
y_pred=model.predict(x_test)
ypred = [outputToLabel(y_pred[i],len(y_test[i])) for i in range(len(y_pred))]
evaluation_report(y_test, ypred)

tag precision   recall  f_score correct_count
0                1   99.8911  99.5929  99.7418          3670
1                2   94.1198  92.9559  93.5342          7667
2                3   89.9381  96.3763   93.046         16277
3                4   99.9689  99.5898   99.779         12869
4                5   79.4521  86.5672  82.8571            58
5                6       100  86.9732  93.0328           454
6                7   98.0948  96.5849   97.334          2008
7                8   69.1943  35.1807  46.6454           146
8                9   73.9777  54.0761  62.4804           199
9               10   59.6078  36.2336  45.0704           304
10              11   92.3077  97.6744  94.9153            84
11              12   94.9214  98.0025  96.4373           785
12              13    85.746  86.0323  85.8889          3092
13              14   95.3909  93.1317  94.2478          5112
14              15   77.2994    70.41   73.694           790
15              16   86.2823  90.3414  88.2652          2170
16              17   95.8188  94.1781  94.9914           550
17              18   97.2743  98.8745  98.0678          1142
18              19   95.3488  95.0725  95.2104           328
19              20   98.2818  96.9492  97.6109           286
20              21   93.2796  91.6172  92.4409          1388
21              22    84.136   74.111   78.806          1188
22              23   89.4667  95.1099   92.202          1342
23              24   82.1235  82.8415   82.481           758
24              25   85.1003  71.9128  77.9528           297
25              26   90.2299  89.2045  89.7143           157
26              27   89.2857  76.3359  82.3045           100
27              29   92.2388  97.7848  94.9309           309
28              30   63.7168  70.5882  66.9767            72
29              31   74.2574  72.8155  73.5294            75
30              32        76  53.3708  62.7063            95
31              33         -        0        -             0
32              34   86.0465  92.1708  89.0034           518
33              35   83.3333  55.5556  66.6667             5
34              36       100     37.5  54.5455             6
35              37   93.4579  98.0392  95.6938           100
36              38        60  30.7692   40.678            12
37              39   67.0157  91.4286  77.3414           128
38              40   99.6441      100  99.8217           280
39              41   66.6667       70  68.2927            14
40              42   92.8571  76.4706   83.871            13
41              43         -        0        -             0
42              45         -        0        -             0
43              46         -        0        -             0
44  accuracy=92.59

In [82]:
%%time
reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.05, patience=3, min_lr=0.001)
model.fit(x_train,y_train,batch_size=256,epochs=5,verbose=1,shuffle=True,validation_split=0.2, callbacks=[reduce_lr])

Train on 14800 samples, validate on 3700 samples
Epoch 1/5
14800/14800 [==============================] - 39s - loss: 44.7803 - acc: 0.9535 - val_loss: 43.5166 - val_acc: 0.9192
Epoch 2/5
14800/14800 [==============================] - 39s - loss: 44.7750 - acc: 0.9542 - val_loss: 43.5185 - val_acc: 0.9203
Epoch 3/5
14800/14800 [==============================] - 39s - loss: 44.7688 - acc: 0.9561 - val_loss: 43.5191 - val_acc: 0.9204
Epoch 4/5
14800/14800 [==============================] - 39s - loss: 44.7628 - acc: 0.9572 - val_loss: 43.5222 - val_acc: 0.9203
Epoch 5/5
14800/14800 [==============================] - 39s - loss: 44.7581 - acc: 0.9586 - val_loss: 43.5196 - val_acc: 0.9204
CPU times: user 8min 55s, sys: 1min 13s, total: 10min 9s
Wall time: 3min 17s


In [83]:
y_pred=model.predict(x_test)
ypred = [outputToLabel(y_pred[i],len(y_test[i])) for i in range(len(y_pred))]
evaluation_report(y_test, ypred)

tag precision   recall  f_score correct_count
0                1   99.8911  99.5929  99.7418          3670
1                2   94.3702  92.4709  93.4109          7627
2                3   90.2278  96.1632   93.101         16241
3                4   99.9689  99.4583  99.7129         12852
4                5   86.9565  89.5522  88.2353            60
5                6   99.7802  86.9732  92.9376           454
6                7   98.3831  96.5849  97.4757          2008
7                8   64.5038  40.7229  49.9261           169
8                9   73.8019  62.7717  67.8414           231
9               10    58.498  35.2801  44.0149           296
10              11   89.3617  97.6744  93.3333            84
11              12   96.1395  96.3795  96.2594           772
12              13   85.3047  86.0879  85.6945          3094
13              14   94.9807  94.1155  94.5461          5166
14              15   79.7392  70.8556  75.0354           795
15              16   87.1046  89.4255  88.2498          2148
16              17   96.3532  85.9589  90.8597           502
17              18   96.7824   98.961  97.8596          1143
18              19   97.0674   95.942  96.5015           331
19              20   98.6207  96.9492  97.7778           286
20              21    94.042  91.6832  92.8476          1389
21              22   83.2224  77.9788  80.5153          1250
22              23   87.5809  95.9603  91.5793          1354
23              24   85.6322  81.4208  83.4734           745
24              25   88.1844   74.092  80.5263           306
25              26   91.9075  90.3409  91.1175           159
26              27   90.5172  80.1527  85.0202           105
27              29   92.4471  96.8354  94.5904           306
28              30   71.8447   72.549  72.1951            74
29              31   49.0446  74.7573  59.2308            77
30              32   72.9927  56.1798  63.4921           100
31              33         -        0        -             0
32              34    85.173  91.9929  88.4517           517
33              35   71.4286  55.5556     62.5             5
34              36       100     87.5  93.3333            14
35              37   92.5926  98.0392  95.2381           100
36              38   65.5172  48.7179  55.8824            19
37              39   69.3548  92.1429  79.1411           129
38              40   99.6441      100  99.8217           280
39              41   88.2353       75  81.0811            15
40              42       100  76.4706  86.6667            13
41              43         -        0        -             0
42              45         -        0        -             0
43              46         -        0        -             0
44  accuracy=92.64

In [84]:
model.save_weights('/data/viterbi_crf_3.h5')

In [47]:
%%time
reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.05, patience=3, min_lr=0.001)
model.fit(x_train,y_train,batch_size=128,epochs=5,verbose=1,shuffle=True,validation_split=0.2, callbacks=[reduce_lr])

Train on 14800 samples, validate on 3700 samples
Epoch 1/5
14800/14800 [==============================] - 69s - loss: 44.7495 - acc: 0.9609 - val_loss: 43.5045 - val_acc: 0.9210
Epoch 2/5
14800/14800 [==============================] - 70s - loss: 44.7435 - acc: 0.9621 - val_loss: 43.5086 - val_acc: 0.9219
Epoch 3/5
14800/14800 [==============================] - 70s - loss: 44.7376 - acc: 0.9634 - val_loss: 43.5114 - val_acc: 0.9199
Epoch 4/5
14800/14800 [==============================] - 70s - loss: 44.7331 - acc: 0.9651 - val_loss: 43.5122 - val_acc: 0.9202
Epoch 5/5
14800/14800 [==============================] - 71s - loss: 44.7269 - acc: 0.9666 - val_loss: 43.5234 - val_acc: 0.9200
CPU times: user 15min 46s, sys: 2min 20s, total: 18min 7s
Wall time: 5min 52s


In [48]:
y_pred=model.predict(x_test)
ypred = [outputToLabel(y_pred[i],len(y_test[i])) for i in range(len(y_pred))]
evaluation_report(y_test, ypred)

tag precision   recall  f_score correct_count
0                1    99.891  99.5115  99.7009          3667
1                2      93.8  93.9137  93.8568          7746
2                3   90.8744  94.3395  92.5745         15933
3                4   99.9689  99.3809   99.674         12842
4                5   83.5443  98.5075   90.411            66
5                6   96.8553  88.5057  92.4925           462
6                7   98.0507  96.7773  97.4098          2012
7                8   63.1737  50.8434  56.3418           211
8                9   63.5838  59.7826  61.6246           220
9               10   59.2453  37.4255  45.8729           314
10              11   89.3617  97.6744  93.3333            84
11              12   92.2353  97.8777  94.9727           784
12              13   87.5036  85.1419  86.3066          3060
13              14   93.8144  94.4981   94.155          5187
14              15   72.2272  71.3904  71.8064           801
15              16   88.1552  88.9259  88.5389          2136
16              17   95.9781  89.8973  92.8382           525
17              18    96.228  99.3939  97.7853          1148
18              19    96.793  96.2319  96.5116           332
19              20    98.951  95.9322  97.4182           283
20              21   93.0446  93.5974  93.3202          1418
21              22   77.8313  80.5989  79.1909          1292
22              23   90.2093  94.6846  92.3928          1336
23              24   91.0141  77.4863  83.7072           709
24              25   94.4056  65.3753  77.2532           270
25              26   92.4855  90.9091  91.6905           160
26              27   91.7431  76.3359  83.3333           100
27              29   93.0159  92.7215  92.8685           293
28              30   79.1667  74.5098  76.7677            76
29              31   60.2837  82.5243  69.6721            85
30              32   58.1006   58.427  58.2633           104
31              33       100  4.41176   8.4507             3
32              34   82.1086  91.4591   86.532           514
33              35   55.5556  55.5556  55.5556             5
34              36       100    81.25  89.6552            13
35              37   89.3805  99.0196  93.9535           101
36              38   48.7179  48.7179  48.7179            19
37              39   73.3696  96.4286  83.3333           135
38              40       100      100      100           280
39              41        75       75       75            15
40              42       100  64.7059  78.5714            11
41              43         -        0        -             0
42              45         -        0        -             0
43              46         -        0        -             0
44  accuracy=92.45

In [51]:
%%time
reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.05, patience=3, min_lr=0.001)
model.fit(x_train,y_train,batch_size=128,epochs=2,verbose=1,shuffle=True,validation_split=0.3, callbacks=[reduce_lr])

Train on 12950 samples, validate on 5550 samples
Epoch 1/2
12950/12950 [==============================] - 66s - loss: 45.1413 - acc: 0.9705 - val_loss: 42.9347 - val_acc: 0.9363
Epoch 2/2
12950/12950 [==============================] - 66s - loss: 45.1379 - acc: 0.9717 - val_loss: 42.9397 - val_acc: 0.9360
CPU times: user 5min 54s, sys: 52.4 s, total: 6min 47s
Wall time: 2min 12s


In [52]:
y_pred=model.predict(x_test)
ypred = [outputToLabel(y_pred[i],len(y_test[i])) for i in range(len(y_pred))]
evaluation_report(y_test, ypred)

tag precision   recall  f_score correct_count
0                1   99.8641  99.6744  99.7691          3673
1                2   93.6979  94.0955  93.8963          7761
2                3   90.9699  94.1856  92.5498         15907
3                4     99.93  99.4428  99.6858         12850
4                5   87.1429  91.0448  89.0511            61
5                6   97.4522   87.931  92.4471           459
6                7   97.9116  96.9697  97.4384          2016
7                8   61.5591  55.1807  58.1957           229
8                9   59.6059  65.7609  62.5323           242
9               10   61.2903  38.4982  47.2914           323
10              11   90.2174  96.5116  93.2584            83
11              12    93.599  96.7541  95.1504           775
12              13   87.2582  85.3645   86.301          3068
13              14   94.3207  94.0973  94.2088          5165
14              15   67.0455  73.6185  70.1784           826
15              16   87.5509   89.592  88.5597          2152
16              17   97.9008  87.8425  92.5993           513
17              18   96.3805  99.1342  97.7379          1145
18              19   96.4706  95.0725  95.7664           328
19              20   98.9286  93.8983  96.3478           277
20              21   93.3245  93.2013  93.2629          1412
21              22   81.1271  77.2302  79.1307          1238
22              23   89.1076  96.2438  92.5383          1358
23              24    90.736  78.1421  83.9695           715
24              25   92.0962   64.891  76.1364           268
25              26   95.8824  92.6136  94.2197           163
26              27   94.2308  74.8092  83.4043            98
27              29   93.5185  95.8861  94.6875           303
28              30   76.7677  74.5098  75.6219            76
29              31   56.2914  82.5243  66.9291            85
30              32   64.3836   52.809  58.0247            94
31              33       100  4.41176   8.4507             3
32              34   85.8108  90.3915  88.0416           508
33              35   55.5556  55.5556  55.5556             5
34              36       100    81.25  89.6552            13
35              37   90.9091  98.0392  94.3396           100
36              38   52.7778  48.7179  50.6667            19
37              39   74.0331  95.7143  83.4891           134
38              40       100      100      100           280
39              41   76.1905       80  78.0488            16
40              42       100  64.7059  78.5714            11
41              43         0        0        -             0
42              45         -        0        -             0
43              46         -        0        -             0
44  accuracy=92.45

## 4.2.2 CRF Marginal

### #TODO 5

Use Keras-contrib CRF layer in your model. You should set the layer parameter so it can give the best performance on testing using <b>marginal problabilities</b>. You <b>must not train the model</b> from scratch but use the pretrained weight from previous CRF Viterbi model.

To finish this excercise you must train the model and show the evaluation report with this model as shown in the example.

In [64]:
# Write your code here
K.clear_session()
model = Sequential()
model.add(Embedding(len(word_to_idx),32,input_length=102,mask_zero=True))
model.add(Bidirectional(GRU(32, return_sequences=True)))
model.add(Dropout(0.2))
model.add(TimeDistributed(Dense(n_classes, activation = 'relu')))
crf = CRF(n_classes,
          learn_mode='marginal',
          test_mode='marginal',
          sparse_target=False,
          use_boundary=True,
          use_bias=True,
          activation='linear',
          kernel_initializer='glorot_uniform',
          chain_initializer='orthogonal',
          bias_initializer='zeros',
          boundary_initializer='zeros',
          kernel_regularizer=None,
          chain_regularizer=None,
          boundary_regularizer=None,
          bias_regularizer=None,
          kernel_constraint=None,
          chain_constraint=None,
          boundary_constraint=None,
          bias_constraint=None,
          input_dim=None,
          unroll=False)
model.add(crf)
model.summary()
adam  = Adam(lr=0.0015)
model.compile(optimizer=adam,loss=crf.loss_function, metrics=[crf.accuracy])

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, 102, 32)           480608    
_________________________________________________________________
bidirectional_1 (Bidirection (None, 102, 64)           12480     
_________________________________________________________________
dropout_1 (Dropout)          (None, 102, 64)           0         
_________________________________________________________________
time_distributed_1 (TimeDist (None, 102, 48)           3120      
_________________________________________________________________
crf_1 (CRF)                  (None, 102, 48)           4752      
Total params: 500,960
Trainable params: 500,960
Non-trainable params: 0
_________________________________________________________________


In [65]:
model.fit(x_train,y_train,batch_size=256,epochs=5,verbose=1,shuffle=True,validation_split=0.3, callbacks=[reduce_lr])

Train on 12950 samples, validate on 5550 samples
Epoch 1/5
12950/12950 [==============================] - 41s - loss: 3.0994 - acc: 0.2890 - val_loss: 2.3322 - val_acc: 0.3539
Epoch 2/5
12950/12950 [==============================] - 41s - loss: 1.7014 - acc: 0.5549 - val_loss: 1.1082 - val_acc: 0.7399
Epoch 3/5
12950/12950 [==============================] - 41s - loss: 0.6938 - acc: 0.8325 - val_loss: 0.5602 - val_acc: 0.8738
Epoch 4/5
12950/12950 [==============================] - 41s - loss: 0.3783 - acc: 0.9089 - val_loss: 0.4497 - val_acc: 0.8864
Epoch 5/5
12950/12950 [==============================] - 41s - loss: 0.2823 - acc: 0.9311 - val_loss: 0.4058 - val_acc: 0.8936


In [66]:
y_pred=model.predict(x_test)
ypred = [outputToLabel(y_pred[i],len(y_test[i])) for i in range(len(y_pred))]
evaluation_report(y_test, ypred)

tag precision   recall  f_score correct_count
0                1   99.8911  99.5929  99.7418          3670
1                2   94.7627  91.2585  92.9776          7527
2                3    88.502  96.1632  92.1737         16241
3                4   99.9766  99.3964  99.6857         12844
4                5    95.082  86.5672   90.625            58
5                6   99.1266  86.9732  92.6531           454
6                7   96.5567   97.114  96.8345          2019
7                8   78.1818  31.0843  44.4828           129
8                9   72.4771  21.4674  33.1237            79
9               10    54.159  34.9225  42.4638           293
10              11   91.7647  90.6977  91.2281            78
11              12   92.8307  93.7578  93.2919           751
12              13   80.5578  84.3907  82.4297          3033
13              14   94.6674  91.5285  93.0715          5024
14              15   71.7414  61.3191  66.1221           688
15              16   82.7175  89.4671    85.96          2149
16              17   96.0573  91.7808  93.8704           536
17              18   96.4527  98.8745  97.6486          1142
18              19   97.9104  95.0725  96.4706           328
19              20   97.9452  96.9492  97.4446           286
20              21   90.9333   90.033  90.4809          1364
21              22   74.1772  73.1129  73.6412          1172
22              23   83.4155  95.8894  89.2186          1353
23              24    76.332  81.4208  78.7943           745
24              25   79.2727  52.7845  63.3721           218
25              26        88       75  80.9816           132
26              27   89.1304  62.5954  73.5426            82
27              29   94.3396  94.9367  94.6372           300
28              30   66.9903  67.6471  67.3171            69
29              31   63.6364  81.5534  71.4894            84
30              32         -        0        -             0
31              33         -        0        -             0
32              34   90.6087  92.7046  91.6447           521
33              35         -        0        -             0
34              36         -        0        -             0
35              37   78.2946  99.0196  87.4459           101
36              38         -        0        -             0
37              39   58.3815  72.1429  64.5367           101
38              40   99.6441      100  99.8217           280
39              41         -        0        -             0
40              42         -        0        -             0
41              43         -        0        -             0
42              45         -        0        -             0
43              46         -        0        -             0
44  accuracy=91.17

In [67]:
model.fit(x_train,y_train,batch_size=256,epochs=5,verbose=1,shuffle=True,validation_split=0.3, callbacks=[reduce_lr])

Train on 12950 samples, validate on 5550 samples
Epoch 1/5
12950/12950 [==============================] - 41s - loss: 0.2387 - acc: 0.9387 - val_loss: 0.3884 - val_acc: 0.8949
Epoch 2/5
12950/12950 [==============================] - 41s - loss: 0.2132 - acc: 0.9430 - val_loss: 0.3679 - val_acc: 0.9016
Epoch 3/5
12950/12950 [==============================] - 41s - loss: 0.1974 - acc: 0.9454 - val_loss: 0.3536 - val_acc: 0.9092
Epoch 4/5
12950/12950 [==============================] - 41s - loss: 0.1845 - acc: 0.9479 - val_loss: 0.3416 - val_acc: 0.9150
Epoch 5/5
12950/12950 [==============================] - 41s - loss: 0.1751 - acc: 0.9501 - val_loss: 0.3337 - val_acc: 0.9177


In [68]:
y_pred=model.predict(x_test)
ypred = [outputToLabel(y_pred[i],len(y_test[i])) for i in range(len(y_pred))]
evaluation_report(y_test, ypred)

tag precision   recall  f_score correct_count
0                1   99.8911  99.5929  99.7418          3670
1                2   95.3231  91.9253  93.5934          7582
2                3   89.4036  96.2165   92.685         16250
3                4   99.9689  99.6053  99.7868         12871
4                5   95.2381  89.5522  92.3077            60
5                6   99.1266  86.9732  92.6531           454
6                7   97.4818  96.8254  97.1525          2013
7                8   74.1935  38.7952  50.9494           161
8                9   73.1602  45.9239  56.4274           169
9               10   64.1618  39.6901  49.0427           333
10              11   93.2584  96.5116  94.8571            83
11              12   95.0739  96.3795  95.7223           772
12              13   84.7453  85.1697   84.957          3061
13              14   94.3965  93.6054  93.9993          5138
14              15   77.5553  65.5971  71.0768           736
15              16   86.9337  89.4671  88.1822          2149
16              17   97.5104  80.4795  88.1801           470
17              18   96.3805  99.1342  97.7379          1145
18              19   97.3294  95.0725  96.1877           328
19              20   97.9452  96.9492  97.4446           286
20              21    92.074  92.0132  92.0436          1394
21              22   80.6706   76.544  78.5531          1227
22              23   86.3752  96.5982  91.2011          1363
23              24   80.6277  81.4208  81.0223           745
24              25   86.8167  65.3753  74.5856           270
25              26   91.3295  89.7727  90.5444           158
26              27   92.7273  77.8626  84.6473           102
27              29   92.5373  98.1013  95.2381           310
28              30   70.4762   72.549  71.4976            74
29              31   43.5897  82.5243   57.047            85
30              32   78.8618  54.4944  64.4518            97
31              33         -        0        -             0
32              34   88.8889  91.1032  89.9824           512
33              35   55.5556  55.5556  55.5556             5
34              36         -        0        -             0
35              37      80.8  99.0196  88.9868           101
36              38      32.5  33.3333  32.9114            13
37              39   69.3182  87.1429  77.2152           122
38              40   98.5915      100  99.2908           280
39              41        70       70       70            14
40              42       100  64.7059  78.5714            11
41              43         -        0        -             0
42              45         -        0        -             0
43              46         -        0        -             0
44  accuracy=92.26

In [69]:
model.fit(x_train,y_train,batch_size=256,epochs=5,verbose=1,shuffle=True,validation_split=0.3, callbacks=[reduce_lr])

Train on 12950 samples, validate on 5550 samples
Epoch 1/5
12950/12950 [==============================] - 40s - loss: 0.1658 - acc: 0.9518 - val_loss: 0.3321 - val_acc: 0.9176
Epoch 2/5
12950/12950 [==============================] - 40s - loss: 0.1586 - acc: 0.9538 - val_loss: 0.3263 - val_acc: 0.9195
Epoch 3/5
12950/12950 [==============================] - 40s - loss: 0.1531 - acc: 0.9547 - val_loss: 0.3230 - val_acc: 0.9201
Epoch 4/5
12950/12950 [==============================] - 41s - loss: 0.1470 - acc: 0.9562 - val_loss: 0.3201 - val_acc: 0.9211
Epoch 5/5
12950/12950 [==============================] - 40s - loss: 0.1415 - acc: 0.9577 - val_loss: 0.3207 - val_acc: 0.9224


In [70]:
y_pred=model.predict(x_test)
ypred = [outputToLabel(y_pred[i],len(y_test[i])) for i in range(len(y_pred))]
evaluation_report(y_test, ypred)

tag precision   recall  f_score correct_count
0                1   99.8097  99.6472  99.7284          3672
1                2   94.8731  92.8831  93.8675          7661
2                3   89.3545  96.9625  93.0032         16376
3                4   99.9689  99.5434  99.7557         12863
4                5   95.2381  89.5522  92.3077            60
5                6   99.7802  86.9732  92.9376           454
6                7    96.887  97.3064  97.0962          2023
7                8   73.5632  46.2651  56.8047           192
8                9   75.2941  52.1739  61.6372           192
9               10   66.3286   38.975  49.0991           327
10              11      87.5  97.6744  92.3077            84
11              12    94.813  98.1273  96.4417           786
12              13   89.1588  83.7507  86.3702          3010
13              14   95.2434  93.3868   94.306          5126
14              15   78.9631  70.5882  74.5412           792
15              16   87.5557  89.9251  88.7246          2160
16              17   97.3196  80.8219  88.3068           472
17              18   96.3835  99.2208  97.7816          1146
18              19    97.076  96.2319  96.6521           332
19              20   98.9619  96.9492  97.9452           286
20              21   92.5099  92.9373  92.7231          1408
21              22   83.2888  77.7293   80.413          1246
22              23   88.2924  96.7399  92.3233          1365
23              24   85.4525  81.5301  83.4452           746
24              25   88.9251  66.1017  75.8333           273
25              26   84.9741  93.1818  88.8889           164
26              27   93.1373  72.5191  81.5451            95
27              29   93.7888  95.5696  94.6708           302
28              30    72.381  74.5098    73.43            76
29              31   45.8333  85.4369   59.661            88
30              32   63.9456   52.809  57.8462            94
31              33         -        0        -             0
32              34    88.676  90.5694  89.6127           509
33              35   55.5556  55.5556  55.5556             5
34              36       100     87.5  93.3333            14
35              37   87.8261  99.0196  93.0876           101
36              38   45.4545  38.4615  41.6667            15
37              39   73.3766  80.7143  76.8707           113
38              40   99.6441      100  99.8217           280
39              41   71.4286       75  73.1707            15
40              42       100  64.7059  78.5714            11
41              43         -        0        -             0
42              45         -        0        -             0
43              46         -        0        -             0
44  accuracy=92.71

In [71]:
model.save_weights('/data/marginal_crf.h5')

### #TODO 6

Please pick the best example that can show the different between CRF that use viterbi and CRF that use marginal problabilities. Compare the result and provide a convincing reason. (which model perform better, why?)

<b>Write your answer here :</b>
<pre style="white-space: wrap; padding-right: 10px">
&nbsp;&nbsp;&nbsp;&nbsp;The result from marginal CRF and viterbi CRF yield similar result. But as seen above the best result is from viterbi CRF with 92.87 accuracy. When compared to marginal probabilities, with 92.71 accuracy, which yield very little difference. `Forward-Backward gives <i>marginal</i> probability for each individual state, <i>Viterbi</i> gives probability of the most likely sequence of states` (<a href="https://stats.stackexchange.com/questions/31746/what-is-the-difference-between-the-forward-backward-and-viterbi-algorithms/222270">reference</a>). Viterbi will give the best result from a given choices while marginal will calculate from inputs to outputs. Viterbi is dynamic programming algorithm, but marginal is greedy algorithm as its core. Because the difference in their nature, DP give best results while greedy give good enough result. Viterbi is prefered if the state is small but marginal will give faster result.
</pre>